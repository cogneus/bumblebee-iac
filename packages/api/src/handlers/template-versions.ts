import {
  apiGatewayWrapper,
  warmupWrapper,
  standardEnvVars,
  PagedItems,
  Token,
} from 'bumblebee-common'
import { ListParams, TemplateVersion, templateVersions } from 'bumblebee-services'
import {
  checkAPIEnvironmentVars,
  checkReadPermissions,
  getClientOptions,
  tryParseValue,
} from '../helpers'
import { TemplatesParams } from '../types'

checkAPIEnvironmentVars()

export const main = warmupWrapper(
  apiGatewayWrapper<ListParams, TemplatesParams<ListParams>, Token, PagedItems<TemplateVersion>>({
    handler: ({
      payload,
      params: { id, sort, cursor, limit, fields, includeCurrent },
      serviceContext,
    }) =>
      templateVersions(serviceContext, id!, {
        sort,
        cursor,
        limit: tryParseValue<number | undefined>(limit, 'number'),
        fields,
        includeCurrent: tryParseValue<boolean | undefined>(includeCurrent, 'boolean'),
        ...payload,
      }),
    ...standardEnvVars(),
    serviceName: 'template-api',
    logName: 'template-versions',
    getClientOptions,
    authCheck: checkReadPermissions,
  })
)
