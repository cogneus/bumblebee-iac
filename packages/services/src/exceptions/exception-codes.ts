import { ExceptionCodes } from 'bumblebee-common'

export const TemplateExceptionCodes = {
  ...ExceptionCodes,
  TemplateNotFound: 'TemplateNotFound',
}
