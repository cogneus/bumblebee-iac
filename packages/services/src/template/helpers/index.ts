export * from './keys.helper'
export * from './validate-template.helper'
export * from './generate-id'
export * from './filter-input.class'
