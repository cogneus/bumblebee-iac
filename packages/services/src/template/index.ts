export * from './add'
export * from './get'
export * from './query'
export * from './remove'
export * from './update'
export * from './versions'
