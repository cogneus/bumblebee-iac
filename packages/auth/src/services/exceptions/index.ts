import { StatusCodes, AppException } from 'bumblebee-common'
import { AuthExceptionCodes } from './exception-codes'

export const KeyNotFound = (key: string) =>
  new AppException({
    code: AuthExceptionCodes.KeyNotFound,
    status: StatusCodes.Forbidden,
    message: `Key with name '${key}' not found`,
    source: KeyNotFound,
  })
