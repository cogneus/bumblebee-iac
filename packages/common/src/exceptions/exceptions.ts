import { AppException } from './app-exception.class'
import { ExceptionCodes } from './exception-codes'
import { StatusCodes } from './status-codes'

export const InvalidToken = () =>
  new AppException({
    code: ExceptionCodes.InvalidToken,
    status: StatusCodes.Forbidden,
    message: 'Forbidden',
    source: InvalidToken,
  })

export const InvalidAuthorizartion = () =>
  new AppException({
    code: ExceptionCodes.InvalidAuthorizartion,
    status: StatusCodes.Forbidden,
    message: 'Forbidden',
    source: InvalidAuthorizartion,
  })

export const EmptyRequest = () =>
  new AppException({
    code: ExceptionCodes.EmptyRequest,
    status: StatusCodes.BadRequest,
    message: 'Request data is missing',
    source: EmptyRequest,
  })

export const InvalidService = (service: string) =>
  new AppException({
    code: ExceptionCodes.InvalidService,
    status: StatusCodes.Error,
    message: `Invalid service. ${service}`,
    source: InvalidService,
  })

export const InvalidRequest = (innerMessage: string) =>
  new AppException({
    code: ExceptionCodes.InvalidRequest,
    status: StatusCodes.BadRequest,
    message: `Request data is invalid. ${innerMessage}`,
    source: InvalidRequest,
  })

export const MissingAttributes = (keys: string[]) =>
  new AppException({
    code: ExceptionCodes.MissingAttributes,
    status: StatusCodes.BadRequest,
    message: `The following attributes are mandatory ${keys.join()}`,
    source: MissingAttributes,
  })

export const EnvVarNotFound = (keys: string[]) =>
  new AppException({
    code: ExceptionCodes.EnvVarNotFound,
    status: StatusCodes.Error,
    message: `The following environment variables not found ${keys.join()}`,
    source: EnvVarNotFound,
  })

export const InvalidModeSource = (resourcePrefix: string, eventSourceARN: string | undefined) =>
  new AppException({
    code: ExceptionCodes.InvalidModeSource,
    status: StatusCodes.Error,
    message: `Cannot derive stage from event source. eventSourceARN: ${eventSourceARN} RESOURCE_PREFIX: ${resourcePrefix}`,
    source: InvalidModeSource,
  })
