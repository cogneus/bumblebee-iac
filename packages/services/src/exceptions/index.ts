import { StatusCodes, AppException } from 'bumblebee-common'
import { TemplateExceptionCodes } from './exception-codes'

export const TemplateNotFound = (id: string) =>
  new AppException({
    code: TemplateExceptionCodes.TemplateNotFound,
    status: StatusCodes.NotFound,
    message: `Template with id '${id}' not found`,
    source: TemplateNotFound,
  })

export const EmptyRequest = () =>
  new AppException({
    code: TemplateExceptionCodes.EmptyRequest,
    status: StatusCodes.BadRequest,
    message: 'Template request data is missing',
    source: EmptyRequest,
  })

export const MissingAttributes = (keys: string[]) =>
  new AppException({
    code: TemplateExceptionCodes.MissingAttributes,
    status: StatusCodes.BadRequest,
    message: `The following attributes are mandatory ${keys.join()}`,
    source: MissingAttributes,
  })
