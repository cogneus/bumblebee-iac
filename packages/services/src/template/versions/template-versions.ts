import { IDBClient, PagedItems, ServiceContext, ensureClient } from 'bumblebee-common'
import { TemplateVersion, DBTemplateItem, ListParams } from '../../types'
import { removeKeyInfo } from '../helpers'
import { MissingAttributes } from '../../exceptions'
import { createCommands } from './create-commands'
import { getVersionParams } from './get-version-params'

export const templateVersions = async <T = TemplateVersion>(
  context: ServiceContext,
  id: string,
  params?: ListParams
): Promise<PagedItems<T>> => {
  if (!id) {
    throw MissingAttributes(['id'])
  }
  const db = ensureClient<IDBClient>(context, 'db')
  const table = db.tables.templates
  const listParams = getVersionParams(table, params)
  const { cursor, limit } = listParams
  const input = createCommands(id, listParams, table.name)
  const { items, cursor: nextCursor } = await db.query<DBTemplateItem>(input, limit, cursor)

  return {
    cursor: nextCursor,
    items: (listParams.fields?.length ? items : items.map(removeKeyInfo)) as T[],
  }
}
