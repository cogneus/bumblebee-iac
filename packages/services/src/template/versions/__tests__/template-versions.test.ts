import { MissingAttributes } from '../../../exceptions'
import { serviceContext } from '../../__mocks__/service-context.mock'
import { dbTemplateItem, templateVersion } from '../../__mocks__/template.mock'
import { templateVersions } from '../template-versions'

describe('List templates tests', () => {
  beforeEach(() => {
    jest.clearAllMocks()
    jest.useFakeTimers()
    jest.setSystemTime(new Date('2023-01-01T00:00:00Z'))
  })

  const { id } = templateVersion
  test('query template request should match snapshot', async () => {
    let input
    ;(serviceContext.clients.db?.query as jest.Mock).mockImplementationOnce((params) => {
      input = params
      return {
        items: [dbTemplateItem, dbTemplateItem, dbTemplateItem],
        cursor: 'mock-cursor',
      }
    })
    const result = await templateVersions(serviceContext, id, {
      cursor: 'mock-cursor',
      limit: 1000,
      includeCurrent: true,
    })
    expect({
      input,
      result,
    }).toMatchSnapshot()
  })

  test('query template request with default params should match snapshot', async () => {
    let input
    ;(serviceContext.clients.db?.query as jest.Mock).mockImplementationOnce((params) => {
      input = params
      return {
        items: [dbTemplateItem, dbTemplateItem, dbTemplateItem],
        cursor: 'mock-cursor',
      }
    })
    const result = await templateVersions(serviceContext, id)
    expect({
      input,
      result,
    }).toMatchSnapshot()
  })

  test('request should throw error if id is null', async () => {
    expect(async () => {
      await templateVersions(serviceContext, null as any)
    }).rejects.toThrow(MissingAttributes(['id']))
  })
})
