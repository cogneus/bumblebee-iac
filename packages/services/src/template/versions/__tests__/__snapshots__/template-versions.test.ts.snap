// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List templates tests query template request should match snapshot 1`] = `
{
  "input": {
    "ExpressionAttributeNames": {
      "#PK": "PK",
    },
    "ExpressionAttributeValues": {
      ":PK": "template#00000000-0000-0000-0000-000000000000",
    },
    "KeyConditionExpression": "(#PK = :PK)",
    "ProjectionExpression": undefined,
    "ScanIndexForward": true,
    "TableName": "name-templates",
  },
  "result": {
    "cursor": "mock-cursor",
    "items": [
      {
        "created": "2023-01-01T00:00:00Z",
        "group": "mock-group",
        "id": "00000000-0000-0000-0000-000000000000",
        "modified": "2023-01-01T00:00:00Z",
        "source": "mock-source",
        "target": "mock-target",
        "template": {},
        "version": 1,
      },
      {
        "created": "2023-01-01T00:00:00Z",
        "group": "mock-group",
        "id": "00000000-0000-0000-0000-000000000000",
        "modified": "2023-01-01T00:00:00Z",
        "source": "mock-source",
        "target": "mock-target",
        "template": {},
        "version": 1,
      },
      {
        "created": "2023-01-01T00:00:00Z",
        "group": "mock-group",
        "id": "00000000-0000-0000-0000-000000000000",
        "modified": "2023-01-01T00:00:00Z",
        "source": "mock-source",
        "target": "mock-target",
        "template": {},
        "version": 1,
      },
    ],
  },
}
`;

exports[`List templates tests query template request with default params should match snapshot 1`] = `
{
  "input": {
    "ExpressionAttributeNames": {
      "#PK": "PK",
      "#SK": "SK",
    },
    "ExpressionAttributeValues": {
      ":PK": "template#00000000-0000-0000-0000-000000000000",
      ":SK": "version",
    },
    "KeyConditionExpression": "(#PK = :PK) and begins_with(#SK, :SK)",
    "ProjectionExpression": undefined,
    "ScanIndexForward": true,
    "TableName": "name-templates",
  },
  "result": {
    "cursor": "mock-cursor",
    "items": [
      {
        "created": "2023-01-01T00:00:00Z",
        "group": "mock-group",
        "id": "00000000-0000-0000-0000-000000000000",
        "modified": "2023-01-01T00:00:00Z",
        "source": "mock-source",
        "target": "mock-target",
        "template": {},
        "version": 1,
      },
      {
        "created": "2023-01-01T00:00:00Z",
        "group": "mock-group",
        "id": "00000000-0000-0000-0000-000000000000",
        "modified": "2023-01-01T00:00:00Z",
        "source": "mock-source",
        "target": "mock-target",
        "template": {},
        "version": 1,
      },
      {
        "created": "2023-01-01T00:00:00Z",
        "group": "mock-group",
        "id": "00000000-0000-0000-0000-000000000000",
        "modified": "2023-01-01T00:00:00Z",
        "source": "mock-source",
        "target": "mock-target",
        "template": {},
        "version": 1,
      },
    ],
  },
}
`;
